extensions/objects/templates/unicode-string-editor.component.spec.ts

```diff
@@ -0,0 +1,164 @@
+    fixture = TestBed.createComponent(UnicodeStringEditorComponent);
+    component = fixture.componentInstance;
+    externalSaveService = TestBed.get(ExternalSaveService);
+    spyOnProperty(externalSaveService, 'onExternalSave').and.returnValue(
+      externalSaveEventEmitter);
+    component.alwaysEditable = false;
+    component.value = 'random value';
+  });
+
+  afterEach(() => {
+    component.ngOnDestroy();
+  });
+
+  it('should initialize component when interaction rule is selected', () => {
+    spyOn(component, 'closeEditor').and.callThrough();
+    spyOn(component.componentSubscriptions, 'add');
+    component.active = true;
+
+    expect(component.active).toBe(true);
+
+    component.ngOnInit();
+
+    // We cannot test what componentSubscriptions.add has been called with since
+    // externalSaveService is a private variable and inside it is an anonymous
+    // fucntion.
+    expect(component.componentSubscriptions.add).toHaveBeenCalled();
+    expect(component.componentSubscriptions.closed).toBe(false);
+    expect(component.alwaysEditable).toBe(false);
+    expect(component.closeEditor).toHaveBeenCalled();
+    expect(component.active).toBe(false);
+  });
+
+  it('should replace value when External Save is emitted', () => {
+    component.active = true;
+    component.ngOnInit();
+    spyOn(component, 'closeEditor').and.callThrough();
+    spyOn(component, 'replaceValue').and.callThrough();
+    component.active = true;
+
+    expect(component.active).toBe(true);
+
+    externalSaveEventEmitter.emit();
+
+    // The this.value cannot be tested for changes since it set before the
+    // emit and will reamin the same after the emit.
+    expect(component.replaceValue).toHaveBeenCalledWith('random value');
+    expect(component.value).toBe('random value');
+    expect(component.closeEditor).toHaveBeenCalled();
+    expect(component.active).toBe(false);
+  });
+
+  it('should replace old value when new value is entered in' +
+  'the text box', () => {
+    expect(component.value).toBe('random value');
+
+    component.updateLocalValue('new random value');
+
+    expect(component.value).toBe('new random value');
+  });
+
+  it('should change active when editor is opened', () => {
+    component.active = false;
+
+    expect(component.active).toBe(false);
+
+    component.openEditor();
+
+    expect(component.active).toBe(true);
+  });
+
+  it('should change active when editor is closed', () => {
+    component.active = true;
+
+    expect(component.active).toBe(true);
+
+    component.closeEditor();
+
+    expect(component.active).toBe(false);
+  });
+
+  it('should replace old value when new value is entered' +
+  'in the input field', () => {
+    expect(component.value).toBe('random value');
+
+    component.replaceValue('new random value');
+
+    expect(component.value).toBe('new random value');
+    expect(component.active).toBe(false);
+  });
+
+  it('should make largeInput true when large input is entered', () => {
```

Comment:
this is just testing 1 happy path for a lot of conditions. Would like to see more tests for that.